CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md

```diff
@@ -0,0 +1,79 @@
+
+System and compiler issues are also discussed where violations of basic floating-point foundations can make it very difficult to implement a robust numerical algorithm.
+For example, the author unequivocally states:
+
+> A language definition that does not require parentheses to be honored is useless for floating-point calculations.
+
+Several compilers and systems over the years have been known to violate this and other necessary basic foundations of floating-point and the IEEE 754 standard, especially when compiler optimizations are turned up.
+In addition, full implementation of the IEEE 754 standard on a system is rare and controlling behavior such as floating-point exceptions, rounding modes, and trap handlers is difficult or impossible in a portable way with many programming languages and libraries.
+This is especially true on some of the more bleeding-edge high-performance computing (HPC) accelerators and systems.
+As a result, numerical algorithm developers often have to rely on a smaller subset of the IEEE standard in order to develop a maximally portable robust implementation.
+
+Finally, the paper contains quite a few proofs of floating-point arithmetic and these proofs serve several purposes.
+First, the proofs establish the foundation of floating-point operations which demonstrate that floating-point is not black magic but is a well established mathematical approximation/discritization of real number arithmetic.
+Second, the proofs show that one can rigorously analyze the expressions in a floating-point algorithm and use that to spot areas of concern.
+Finally, these proofs show how important it is to have a standard like IEEE 754 in designing and implementing numerical algorithms by being able to rely on a few key numerical properties.
+
+In summary, the paper "What Every Computer Scientist Should Know About Floating-Point Arithmetic" represents a must-read for any serious numerical algorithm developer.
+In addition to learning the basics of floating-point, it is important that numerical algorithm developers have a strong knowledge of the IEEE 754 standard and be aware of what properties are supported or not supported on the set of compilers and systems that their algorithm implementations must run.
+And it is especially important to numerical developers exposed to bleeding-edge HPC systems.
+When these early systems violate basic IEEE standards, it is useful for these numerical algorithm developers to provide feedback to the vendors to see if floating-point standards compliance can be restored.
+
+<!---
+Publish: yes 
+Pinned: no
+RSS update: ???
+Topics: Programming languages, High-performance computing (HPC), Reproducibility, Refactoring
+--->
+
+<!---
+LocalWords:  associativity demormalized
```

Comment:
should this be denormalize

Reply:
Thanks!  Addressed in commit bcd4e2f264bb664f80eeaa4be26fbdde6918b46b.